activerecord/lib/active_record/associations/join_dependency.rb

```diff
@@ -114,12 +114,16 @@ def join_constraints(outer_joins, join_type)
           end
         }
 
-        joins.concat outer_joins.flat_map { |oj|
+        joins.concat joins_to_add.flat_map { |oj|
           if join_root.match? oj.join_root
             walk join_root, oj.join_root
           else
             oj.join_root.children.flat_map { |child|
-              make_outer_joins oj.join_root, child
+              if join_type == Arel::Nodes::OuterJoin
+                make_left_outer_joins oj.join_root, child
```

Comment:
This is a good point. Usually we don't start the upgrade guide until before the release but I agree we should warn about this.